posts/services/common.py

```diff
@@ -58,6 +63,45 @@ def add_categories(categories: list[int], post: Post):
     post.save()
 
 
+def update_global_leaderboard_tags(post: Post):
+    # set or update the tags for this post with respect to the global
+    # leaderboard(s) this is a part of
+
+    projects: QuerySet[Project] = post.projects.all()
+
+    # Skip if post is not eligible for global leaderboards
+    if not post.default_project.type == Project.ProjectTypes.SITE_MAIN and not next(
+        (p for p in projects if p.type == Project.ProjectTypes.SITE_MAIN), None
+    ):
+        return
+
+    # Get all global leaderboard dates and create/get corresponding tags
+    to_set_tags = []
+    gl_dates = global_leaderboard_dates()
+    for question in post.get_questions():
+        dates = question.get_global_leaderboard_dates(gl_dates=gl_dates)
+        if dates:
+            tag_name, tag_slug = name_and_slug_for_global_leaderboard_dates(dates)
+            tag, _ = Project.objects.get_or_create(
+                type=Project.ProjectTypes.TAG, name=tag_name, slug=tag_slug, order=1
+            )
+            to_set_tags.append(tag)
+
+    # Update post's global leaderboard tags
+    current_gl_tags = [
+        p
+        for p in projects
+        if p.type == Project.ProjectTypes.TAG
+        and p.name.endswith(GLOBAL_LEADERBOARD_STRING)
+    ]
+    for tag in current_gl_tags:
+        if tag not in to_set_tags:
+            post.projects.remove(tag)
+    for tag in to_set_tags:
+        if tag not in current_gl_tags:
+            post.projects.add(tag)
```

Comment:
I assume calling individual `.remove` and `.add` causes N db requests. I think this could be optimized by gathering final projects list and then calling a single `post.projects.set(*final_list)` then

Reply:
The issue there is `set` will override other project.
This will typically either be 1 or 0 tags to effect, so the optimization won't occur almost ever.
I think this is easier to understand and maintain this way since we won't have to be careful about any non-global leaderboard tags this way.